lib/ts/recipe/session/sessionAuth.tsx

```diff
@@ -31,67 +32,77 @@ type Props = {
-    const buildSessionContext = async (): Promise<SessionContextType> => {
-        const sessionExists = await session.doesSessionExist();
+    const setInitialContext = async (): Promise<void> => {
+        const sessionExists = await doesSessionExist();
 
-        if (!sessionExists) {
-            return {
+        if (sessionExists === false) {
+            return setContext({
                 doesSessionExist: sessionExists,
                 jwtPayload: {},
                 userId: "",
-            };
+            });
         }
 
-        const [jwtPayload, userId] = await Promise.all([session.getJWTPayloadSecurely(), session.getUserId()]);
+        const [jwtPayload, userId] = await Promise.all([getJWTPayloadSecurely(), getUserId()]);
 
-        return {
+        return setContext({
             doesSessionExist: sessionExists,
             jwtPayload,
             userId,
-        };
-    };
-
-    const updateSessionContext = async () => {
-        const sessionContext = await buildSessionContext();
-
-        setSessionContext(sessionContext);
+        });
     };
 
-    const onHandleEvent = async (ctx: OnHandleEventContext) => {
-        switch (ctx.action) {
+    const onHandleEvent = async (event: RecipeEvent) => {
+        switch (event.action) {
+            case "SESSION_CREATED":
+                return setContext({
+                    doesSessionExist: true,
+                    userId: await getUserId(),
+                    jwtPayload: await getJWTPayloadSecurely(),
+                });
             case "REFRESH_SESSION":
+                return setContext({
+                    doesSessionExist: true,
+                    userId: context === undefined ? "" : context.userId,
+                    jwtPayload: await getJWTPayloadSecurely(),
+                });
             case "SIGN_OUT":
             case "UNAUTHORISED":
-                return updateSessionContext();
+                return setContext({
+                    doesSessionExist: false,
+                    userId: "",
+                    jwtPayload: {},
+                });
             default:
                 return;
         }
     };
 
-    // Read the current state
+    // Read and set the current state
     useEffect(() => {
-        updateSessionContext();
+        setInitialContext();
     }, []);
 
-    // Setup listener
+    // Setup listener. This will call addEventListener teardown function when component is unmounted
     useEffect(() => session.addEventListener(onHandleEvent), []);
 
     // If the context is undefined, we are still waiting to know whether session exists.
-    if (sessionContext === undefined) {
+    if (context === undefined) {
         return null;
     }
 
     // If the session doesn't exist and we require auth, redirect to login
-    if (sessionContext.doesSessionExist === false && props.requireAuth === true) {
+    if (context.doesSessionExist === false && props.requireAuth === true) {
         props.redirectToLogin();
```

Comment:
This should be called in `onHandleEvent` or in `setInitialContext`

Reply:
> This should be called in `onHandleEvent`...

Again, I would limit the responsibility of `onHandleEvent` function to just setting the state. Then, take actions based on the state. This makes it a lot clearer, as this is a common pattern :)

> ...or in `setInitialContext`
I disagree. `setInitialContext` should only care about setting context. We may do it in `useEffect` hook next to `setInitialContext`, as a separate function.

This would make it only react to the current state, when the component is rendered, which I assume is not what we want.

Summarizing, I would leave that in render block.

Reply:
Agreed. However, I meant to say that shouldn't a side effect like `props.redirectToLogin();` be called in something like `useEffect` or something other similar function? Cause then it would be like calling `props.redirectToLogin();` instead the `render()` function? 

I am not too familiar with the exact semantics of functional components... so if you think this is fine, then cool!

Reply:
So I think functional components are called multiple times (each render). This means, if this function is called twice very fast when `context.doesSessionExist === false && props.requireAuth === true` is true, you will be calling the user's `redirectToLogin` function twice.. (See [this blog](https://dmitripavlutin.com/react-useeffect-explanation/) - "The component rendering and side-effect logic are independent. So it would be a mistake to perform side-effects directly in the body of the component.")

Since calling `props.redirectToLogin();` depends on `context` and `props`, maybe you could:
```
useEffect(() => {
    // Runs ONCE after initial rendering
    // and after every rendering ONLY IF `prop` or `state` changes
    if (sessionContext.doesSessionExist === false && props.requireAuth === true) {
       props.redirectToLogin();
    }
  }, [prop, context]);
```